File: porting/clean_targets/scripts/compare_features.py
#!/usr/bin/env python3
"""
Compare Python .pt vs Julia .npy feature tensors for parity.

Usage:
    python3 clean_targets/scripts/compare_features.py \
        clean_targets/feature_dumps/python/protein_ligand_ccd_features.pt \
        clean_targets/feature_dumps/julia/protein_ligand_ccd_features/
"""
import os
import sys
import numpy as np
import torch


def load_python_features(pt_path):
    feat = torch.load(pt_path, map_location="cpu", weights_only=False)
    result = {}
    for key in feat:
        val = feat[key]
        if isinstance(val, torch.Tensor):
            result[key] = val.detach().numpy()
        elif isinstance(val, dict):
            sub = {}
            for subkey, subval in val.items():
                if isinstance(subval, torch.Tensor):
                    sub[subkey] = subval.detach().numpy()
            result[key] = sub
    return result


def load_julia_features(npy_dir):
    """Load Julia features from a directory of .npy files."""
    result = {}
    for entry in sorted(os.listdir(npy_dir)):
        full_path = os.path.join(npy_dir, entry)
        if entry.endswith(".npy"):
            key = entry[:-4]
            result[key] = np.load(full_path)
        elif os.path.isdir(full_path):
            # Nested dict (e.g. constraint_feature/)
            sub = {}
            for subentry in sorted(os.listdir(full_path)):
                if subentry.endswith(".npy"):
                    subkey = subentry[:-4]
                    sub[subkey] = np.load(os.path.join(full_path, subentry))
            if sub:
                result[entry] = sub
    return result


def compare_tensor(key, py_val, jl_val, atol=1e-5, subkey=""):
    prefix = f"{key}.{subkey}" if subkey else key

    if not isinstance(py_val, np.ndarray) or not isinstance(jl_val, np.ndarray):
        print(f"  SKIP          {prefix}  (not both arrays)")
        return "skip"

    if py_val.shape != jl_val.shape:
        print(f"  SHAPE MISMATCH  {prefix}:  python={py_val.shape}  julia={jl_val.shape}")
        return "fail"

    if np.issubdtype(py_val.dtype, np.floating) or np.issubdtype(jl_val.dtype, np.floating):
        py_f = py_val.astype(np.float64)
        jl_f = jl_val.astype(np.float64)
        diffs = np.abs(py_f - jl_f)
        maxdiff = float(np.max(diffs))
        n_total = py_f.size
        n_diff = int(np.sum(diffs > atol))

        if maxdiff <= atol:
            print(f"  OK            {prefix:35s}  shape={list(py_val.shape)}  maxdiff={maxdiff:.2e}")
            return "ok"
        else:
            pct = round(100.0 * n_diff / n_total, 1)
            print(f"  MISMATCH      {prefix:35s}  shape={list(py_val.shape)}  maxdiff={maxdiff:.2e}  n_diff={n_diff}/{n_total} ({pct}%)")
            flat_idx = np.argsort(diffs.ravel())[::-1]
            for i in range(min(5, len(flat_idx))):
                idx = np.unravel_index(flat_idx[i], py_f.shape)
                print(f"    {idx}: py={py_f[idx]:.6f}  jl={jl_f[idx]:.6f}  diff={diffs[idx]:.6f}")
            return "fail"
    else:
        py_i = py_val.astype(np.int64)
        jl_i = jl_val.astype(np.int64)
        n_diff = int(np.sum(py_i != jl_i))

        if n_diff == 0:
            print(f"  OK            {prefix:35s}  shape={list(py_val.shape)}  (integer exact)")
            return "ok"
        else:
            n_total = py_i.size
            pct = round(100.0 * n_diff / n_total, 1)
            print(f"  MISMATCH      {prefix:35s}  shape={list(py_val.shape)}  n_diff={n_diff}/{n_total} ({pct}%)")
            diff_mask = py_i != jl_i
            shown = 0
            it = np.nditer(diff_mask, flags=['multi_index'])
            while not it.finished and shown < 10:
                if it[0]:
                    idx = it.multi_index
                    print(f"    {idx}: py={py_i[idx]}  jl={jl_i[idx]}")
                    shown += 1
                it.iternext()
            return "fail"


def main():
    if len(sys.argv) < 3:
        print("Usage: compare_features.py <python.pt> <julia_dir/>")
        sys.exit(1)

    py_path, jl_path = sys.argv[1], sys.argv[2]
    print(f"Loading Python features from: {py_path}")
    py = load_python_features(py_path)
    print(f"Loading Julia features from: {jl_path}")
    jl = load_julia_features(jl_path)

    py_keys = set(py.keys())
    jl_keys = set(jl.keys())

    print(f"\n{'='*70}")
    print(f"Python-only keys: {sorted(py_keys - jl_keys)}")
    print(f"Julia-only keys:  {sorted(jl_keys - py_keys)}")
    print(f"Common keys:      {len(py_keys & jl_keys)}")
    print(f"{'='*70}\n")

    n_ok = n_fail = n_skip = 0

    for key in sorted(py_keys & jl_keys):
        py_val = py[key]
        jl_val = jl[key]

        if isinstance(py_val, dict) and isinstance(jl_val, dict):
            for subkey in sorted(set(py_val.keys()) & set(jl_val.keys())):
                r = compare_tensor(key, py_val[subkey], jl_val[subkey], subkey=subkey)
                if r == "ok": n_ok += 1
                elif r == "fail": n_fail += 1
                else: n_skip += 1
        elif isinstance(py_val, np.ndarray) and isinstance(jl_val, np.ndarray):
            r = compare_tensor(key, py_val, jl_val)
            if r == "ok": n_ok += 1
            elif r == "fail": n_fail += 1
            else: n_skip += 1
        else:
            print(f"  SKIP          {key}  (types differ: py={type(py_val).__name__} jl={type(jl_val).__name__})")
            n_skip += 1

    print(f"\n{'='*70}")
    print(f"Summary: {n_ok} OK, {n_fail} MISMATCH, {n_skip} SKIP")
    print(f"{'='*70}")
    sys.exit(1 if n_fail > 0 else 0)


if __name__ == "__main__":
    main()

File: clean_targets/scripts/compare_features.jl
#!/usr/bin/env julia
#
# Compare Python vs Julia feature tensors for parity.
#
# Usage:
#   cd /home/claudey/FixingKAFA/ka_run_env
#   julia --project=. ../PXDesign.jl/clean_targets/scripts/compare_features.jl \
#       ../PXDesign.jl/clean_targets/feature_dumps/python/protein_ligand_ccd_features.pt \
#       ../PXDesign.jl/clean_targets/feature_dumps/julia/protein_ligand_ccd_features.jld2

using JLD2
using PyCall

const torch = pyimport("torch")
const np = pyimport("numpy")

function load_python_features(pt_path::String)
    py_feat = torch.load(pt_path, map_location="cpu", weights_only=false)
    result = Dict{String, Any}()
    for key in py_feat.keys()
        val = py_feat[key]
        if PyCall.pyisinstance(val, torch.Tensor)
            # Convert to Julia array
            arr = val.detach().numpy()
            result[string(key)] = convert(Array, arr)
        elseif PyCall.pyisinstance(val, PyCall.pybuiltin("dict"))
            sub = Dict{String, Any}()
            for subkey in val.keys()
                subval = val[subkey]
                if PyCall.pyisinstance(subval, torch.Tensor)
                    sub[string(subkey)] = convert(Array, subval.detach().numpy())
                end
            end
            result[string(key)] = sub
        end
    end
    return result
end

function load_julia_features(jld2_path::String)
    data = load(jld2_path)
    return data["feat"]
end

# Python uses row-major (N_tok, C), Julia uses column-major (C, N_tok) for some tensors.
# For comparison, we keep Python layout (features-last) and check Julia features-last too,
# since dump_julia_features.jl dumps the raw dict BEFORE as_protenix_features transposition.

function compare_tensor(key::String, py_val, jl_val; atol=1e-5, subkey="")
    prefix = isempty(subkey) ? key : "$key.$subkey"

    if py_val isa AbstractArray && jl_val isa AbstractArray
        if size(py_val) != size(jl_val)
            println("  SHAPE MISMATCH  $prefix:  python=$(size(py_val))  julia=$(size(jl_val))")
            return false
        end

        if eltype(py_val) <: AbstractFloat || eltype(jl_val) <: AbstractFloat
            py_f = Float64.(py_val)
            jl_f = Float64.(jl_val)
            maxdiff = maximum(abs.(py_f .- jl_f))
            n_total = length(py_f)
            n_diff = count(abs.(py_f .- jl_f) .> atol)

            if maxdiff <= atol
                println("  OK            $prefix  shape=$(size(py_val))  maxdiff=$maxdiff")
                return true
            else
                pct = round(100.0 * n_diff / n_total; digits=1)
                println("  MISMATCH      $prefix  shape=$(size(py_val))  maxdiff=$maxdiff  n_diff=$n_diff/$n_total ($pct%)")

                # Show first few mismatches
                diffs = abs.(py_f .- jl_f)
                worst_idx = sortperm(vec(diffs); rev=true)
                for i in 1:min(5, length(worst_idx))
                    idx = worst_idx[i]
                    ci = CartesianIndices(py_f)[idx]
                    println("    [$ci]: py=$(py_f[idx])  jl=$(jl_f[idx])  diff=$(diffs[idx])")
                end
                return false
            end
        else
            # Integer comparison
            py_i = Int64.(py_val)
            jl_i = Int64.(jl_val)
            n_diff = count(py_i .!= jl_i)
            if n_diff == 0
                println("  OK            $prefix  shape=$(size(py_val))  (integer exact)")
                return true
            else
                n_total = length(py_i)
                pct = round(100.0 * n_diff / n_total; digits=1)
                println("  MISMATCH      $prefix  shape=$(size(py_val))  n_diff=$n_diff/$n_total ($pct%)")

                # Show first few mismatches
                for (k, ci) in enumerate(CartesianIndices(py_i))
                    py_i[ci] == jl_i[ci] && continue
                    println("    [$ci]: py=$(py_i[ci])  jl=$(jl_i[ci])")
                    k >= 10 && break
                end
                return false
            end
        end
    else
        println("  SKIP          $prefix  (not both arrays: py=$(typeof(py_val)) jl=$(typeof(jl_val)))")
        return true
    end
end

function main(py_path::String, jl_path::String)
    println("Loading Python features from: $py_path")
    py = load_python_features(py_path)
    println("Loading Julia features from: $jl_path")
    jl = load_julia_features(jl_path)

    py_keys = Set(keys(py))
    jl_keys = Set(keys(jl))

    println("\n", "="^70)
    println("Python-only keys: ", sort(collect(setdiff(py_keys, jl_keys))))
    println("Julia-only keys:  ", sort(collect(setdiff(jl_keys, py_keys))))
    println("Common keys:      ", length(intersect(py_keys, jl_keys)))
    println("="^70, "\n")

    n_ok = 0
    n_fail = 0
    n_skip = 0

    for key in sort(collect(intersect(py_keys, jl_keys)))
        py_val = py[key]
        jl_val = jl[key]

        if py_val isa Dict && jl_val isa Dict
            # Compare nested dict (constraint_feature)
            sub_py_keys = Set(keys(py_val))
            sub_jl_keys = Set(keys(jl_val))
            for subkey in sort(collect(intersect(sub_py_keys, sub_jl_keys)))
                ok = compare_tensor(key, py_val[subkey], jl_val[subkey]; subkey=subkey)
                if ok; n_ok += 1; else; n_fail += 1; end
            end
        elseif py_val isa AbstractArray && jl_val isa AbstractArray
            ok = compare_tensor(key, py_val, jl_val)
            if ok; n_ok += 1; else; n_fail += 1; end
        else
            println("  SKIP          $key  (types: py=$(typeof(py_val)) jl=$(typeof(jl_val)))")
            n_skip += 1
        end
    end

    println("\n", "="^70)
    println("Summary: $n_ok OK, $n_fail MISMATCH, $n_skip SKIP")
    println("="^70)
    return n_fail
end

if length(ARGS) < 2
    println("Usage: julia compare_features.jl <python.pt> <julia.jld2>")
    exit(1)
end

n_fail = main(ARGS[1], ARGS[2])
exit(n_fail > 0 ? 1 : 0)
